util/pkg/vfs/s3fs.go

```diff
@@ -315,14 +321,16 @@ func (p *S3Path) ReadTree() ([]Path, error) {
 
 func (p *S3Path) client() (*s3.S3, error) {
 	var err error
-	if p.region == "" {
-		p.region, err = p.s3Context.getRegionForBucket(p.bucket)
+	if p.bucketDetails == nil || p.bucketDetails.region == "" {
+		bucketDetails, err := p.s3Context.getDetailsForBucket(p.bucket)
+
+		p.bucketDetails = bucketDetails
```

Comment:
Ah I see what you're doing now returning the bucketDetails always.  It's a slightly hard pattern to reason about though - because a retry will succeed I think.